h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1068,8 +1070,16 @@ public long tryCommit() {
      * @return the new version (incremented if there were changes)
      */
     public synchronized long commit() {
-        currentStoreThread.set(Thread.currentThread());
-        store();
+        Thread currentThread = Thread.currentThread();
+        Thread storeThread = currentStoreThread.get();
+        if (currentThread != storeThread) { // to avoid re-entrance
+            currentStoreThread.set(currentThread);
```

Comment:
If `storeThread == null` concurrent `tryCommit()` or `deregisterVersionUsage()` may change its value between `get()` and `set()`, they are not synchronized like this `commit()` method. It looks like `compareAndSet()` should be used here too instead of `set()` or something else should be changed.

Reply:
Any updates?